### PR TITLE
API-10941 added method list_groups_properties

### PR DESCRIPTION
**Pull request by maciejjwojcik**

### CODE REVIEW

File: configuration/api.go [hunk 1]
```diff
@@ -513,3 +513,14 @@ func (a *API) UpdateTag(name string, groupIDs []int) error {
 		GroupIDs: groupIDs,
 	}, &emptyResponse{})
 }
+
+// Lists properties of groups
+func (a *API) ListGroupsProperties(namespace string, namespacePrefix string, groupIDs []int) ([]*Group, error) {
+	var resp []*Group
```
Comment (grz-gajda) [inline configuration/api.go]: `list_groups_properties` doesn't return `*Group` but its properties (check type `Properties`).

File: configuration/api_test.go [hunk 2]
```diff
@@ -1258,3 +1269,19 @@ func TestUpdateTagShouldReturnDataReceivedFromConfApi(t *testing.T) {
...
+	if err != nil {
+		t.Error("API creation failed")
+	}
+
+	resp, err := api.ListGroupsProperties("namespace", "prefix,", []int{0, 1})
+
+	if err != nil {
+		t.Errorf("ListGroupsProperties failed: %v", err)
+	}
+	fmt.Print(resp)
```
Comment (grz-gajda) [inline configuration/api_test.go]: Remove `fmt`.

File: configuration/types.go [hunk 3]
```diff
@@ -41,6 +41,7 @@ const (
 	BotCreated                   WebhookAction = "bot_created"
 	BotUpdated                   WebhookAction = "bot_updated"
 	BotDeleted                   WebhookAction = "bot_deleted"
+	ListGroupsProperties         WebhookAction = "list_groups_properties"
```
Comment (grz-gajda) [inline configuration/types.go]: There isn't any webhook named `list_groups_properties`.

File: webhooks/handler.go [hunk 4]
```diff
@@ -163,6 +163,8 @@ func NewWebhookHandler(cfg *Configuration) http.HandlerFunc {
 			payload = &BotUpdated{}
 		case "bot_deleted":
 			payload = &BotDeleted{}
+		case "list_groups_properties":
+			payload = &ListGroupsProperties{}
```
Comment (grz-gajda) [inline webhooks/handler.go]: There isn't any webhook named `list_groups_properties`.

File: webhooks/handler_test.go [hunk 5]
```diff
@@ -50,6 +50,7 @@ var verifiers = map[string]webhooks.Handler{
 	"bot_created":                     botCreated,
 	"bot_updated":                     botUpdated,
 	"bot_deleted":                     botDeleted,
+	"list_groups_properties":          listGroupsProperties,
```
Comment (grz-gajda) [inline webhooks/handler_test.go]: There isn't any webhook named `list_groups_properties`.

File: webhooks/helpers_test.go [hunk 6]
```diff
@@ -788,3 +788,19 @@ func autoAccessDeleted(ctx context.Context, wh *webhooks.Webhook) error {
...
+	var propEqErrors string
+	propEq("ID", payload.ID, 42, &propEqErrors)
+
+	propEq("Properties.Namespace.Name.Value", payload.Properties["namespace"]["name"], "sales", &propEqErrors)
+
+	if propEqErrors != "" {
+		return errors.New(propEqErrors)
+	}
+	return nil
+}
```
Comment (grz-gajda) [inline webhooks/helpers_test.go]: There isn't any webhook named `list_groups_properties`.

File: webhooks/structures.go [hunk 7]
```diff
@@ -276,6 +276,12 @@ type BotDeleted struct {
 	ID string `json:"id"`
 }
 
+// ListGroupsProperties represents payload of list_groups_properties webhook.
+type ListGroupsProperties struct {
+	ID         int        `json:"id"`
+	Properties Properties `json:"properties"`
+}
+
```
Comment (grz-gajda) [inline webhooks/structures.go]: There isn't any webhook named `list_groups_properties`.

File: webhooks/testdata/list_groups_properties.json [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+	"payload": {
+		"id": 42,
+		"properties":{
+			"namespace":{
+				"name": "sales"
+			}
+		}
+	},
+	"additional_data": {}
+}
```
Comment (grz-gajda) [inline webhooks/testdata/list_groups_properties.json]: There isn't any webhook named `list_groups_properties`.

File: configuration/api_test.go [hunk 9]
```diff
@@ -310,6 +311,16 @@ var mockedResponses = map[string]string{
...
 	"update_tag": `{}`,
+	"list_groups_properties": `[
+		{
+		  "id": 0,
+		  "properties": {
+			"abc": {
+			  "a_property": "a"
+			}
+		  }
+		}
```
Comment (grz-gajda) [inline configuration/api_test.go]: Adjust intending to one length.

**Review by grz-gajda — CHANGES_REQUESTED** [review]

File: configuration/api.go (hunk 1 above)
Comment (maciejjwojcik) [reply configuration/api.go]: Done.

File: configuration/api_test.go (hunk 2 above)
Comment (maciejjwojcik) [reply configuration/api_test.go]: Removed.

File: configuration/types.go (hunk 3 above)
Comment (maciejjwojcik) [reply configuration/types.go]: Removed any indication list_groups_properties is a hook.

File: configuration/api_test.go (hunk 9 above)
Comment (maciejjwojcik) [reply configuration/api_test.go]: Fixed indenting.

File: configuration/api.go [hunk 10]
```diff
@@ -513,3 +513,14 @@ func (a *API) UpdateTag(name string, groupIDs []int) error {
 		GroupIDs: groupIDs,
 	}, &emptyResponse{})
 }
+
+// Lists properties of groups
+func (a *API) ListGroupsProperties(namespace string, namespacePrefix string, groupIDs []int) ([]listGroupsPropertiesResponse, error) {
```
Comment (grz-gajda) [inline configuration/api.go]: We shouldn't return private `struct` to the developers. I think you could create `GroupProperties`:

```go
type GroupProperties struct {
	ID         int32
	Properties Properties
}
```

Comment (grz-gajda) [reply configuration/api.go]: Also, more often will be used to property `groupIDs` than `namespace`. I suggest to reorder fields: `groupIDs` -> `namespace` -> `namePrefix` (it's not `namespacePrefix`).

**Review by grz-gajda — CHANGES_REQUESTED** [review]

File: configuration/api.go (hunk 10 above)
Comment (maciejjwojcik) [reply configuration/api.go]: Fixed acc. to suggestions.

File: configuration/internal.go [hunk 11]
```diff
@@ -234,3 +234,9 @@ type updateTagRequest struct {
 	Name     string `json:"name"`
 	GroupIDs []int  `json:"group_ids"`
 }
+
+type listGroupsPropertiesRequest struct {
+	GroupIDs   []int  `json:"group_ids,omitempty"`
+	Namespace  string `json:"namespace,omitempty"`
+	NamePrefix string `json:"namespace_prefix,omitempty"`
```
Comment (grz-gajda) [inline configuration/internal.go]: Copied from internal codebase:

```go
type ListGroupsPropertiesRequest struct {
	NamePrefix      string  `json:"name_prefix"`
}
```

you need to be consistent with fields naming.

**Review by grz-gajda — CHANGES_REQUESTED** [review]

File: configuration/internal.go (hunk 11 above)
Comment (maciejjwojcik) [reply configuration/internal.go]: Fixed. Also double-checked for other instances of namespace_prefix introduced by me, but none was found.

**Review by grz-gajda — APPROVED** [review]

**Pull request merged**